src/macaron/repo_finder/repo_verifier.py

```diff
@@ -0,0 +1,300 @@
+
+    status: RepositoryVerificationStatus
+    reason: str
+    build_tool: BaseBuildTool
+
+
+class RepoVerifier:
+    """A class to verify a claimed repo."""
+
+    _known_maven_namespaces = {
+        "github",
+        "gitlab",
+        "bitbucket",
+        "gitee",
+    }
+
+    def __init__(
+        self,
+        namespace: str,
+        name: str,
+        version: str,
+        claimed_repo_url: str,
+        claimed_repo_fs: str,
+        build_tool: BaseBuildTool,
+    ):
+        self.namespace = namespace
+        self.name = name
+        self.version = version
+        self.claimed_repo_url = claimed_repo_url
+        self.claimed_repo_fs = claimed_repo_fs
+        self.build_tool = build_tool
+
+    def verify_claimed_repo(self) -> RepositoryVerificationResult:
+        """Verify whether the claimed repository links back to the artifact."""
+        default_res = RepositoryVerificationResult(
+            status=RepositoryVerificationStatus.UNKNOWN, reason="unsupported_type", build_tool=self.build_tool
+        )
+
+        match self.build_tool:
+            case Maven():
+                git_ns_ver = self._verify_maven_git_ns()
+                if git_ns_ver.status == RepositoryVerificationStatus.PASSED:
+                    return git_ns_ver
+                return self._verify_maven()
+            case Gradle():
+                git_ns_ver = self._verify_maven_git_ns()
+                if git_ns_ver.status == RepositoryVerificationStatus.PASSED:
+                    return git_ns_ver
+                return self._verify_gradle()
+            # TODO: add verifiers for other types
+            case Poetry():
+                return default_res
+            case Pip():
+                return default_res
+            case Docker():
+                return default_res
+            case NPM():
+                return default_res
+            case Yarn():
+                return default_res
+            case Go():
+                return default_res
+            case _:
+                raise NotImplementedError(f"Unsupported build tool: {self.build_tool}")
+
+    def _same_group(self, g1: str, g2: str) -> bool:
+        if g1 == g2:
+            return True
+
+        g1_parts = g1.split(".")
+        g2_parts = g2.split(".")
+        if min(len(g1_parts), len(g2_parts)) < 2:
+            return False
+
+        if (g1_parts[0] in {"io", "com"} and g1_parts[1] in self._known_maven_namespaces) or (
+            g2_parts[0] in {"io", "com"} and g2_parts[1] in self._known_maven_namespaces
+        ):
+            if len(g1_parts) >= 3 and len(g2_parts) >= 3:
+                return all(g1_parts[i] == g2_parts[i] for i in range(3))
+            return False
+
+        for i in range(2):
+            if g1_parts[i] != g2_parts[i]:
+                return False
+
+        return True
+
+    @staticmethod
+    def _bfs_walk(root_dir: Path, filename: str) -> Path | None:
```

Comment:
You can probably use the [`get_build_dirs`](https://github.com/oracle/macaron/blob/edfe06eb3f9337f86215b253f40f53ca1a0105c9/src/macaron/slsa_analyzer/build_tool/base_build_tool.py#L190) function in build tools. You might just need to improve it to avoid looking into the `["test", "example", "sample", "doc", "demo", "spec", "mock"]` directories as implemented in this function.

Reply:
Interesting. Yes, I that's the correct funciton to be used here. Thanks!

Reply:
I decided to keep this for now (with another name `find_file_in_repo` under `repo_verifier_base`) because there amount of refactoring needed to make use of those functions for this purpose seems to be quite high.

Reply:
OK, no worries. Can you please add a TODO comment to refactor and use [get_build_dirs](https://github.com/oracle/macaron/blob/edfe06eb3f9337f86215b253f40f53ca1a0105c9/src/macaron/slsa_analyzer/build_tool/base_build_tool.py#L190) later?